### DISCUSSION TITLE
libvips with Imagemagick in production

### DISCUSSION THREAD

**Discussion by michielswaanen** [libvips/libvips#3676] [Q&A]
Hi everyone,

I'm building a web server that can convert images to other image formats. I want to expand the initial set of formats and add support for EPS files. After doing some research, I found that both ImageMagick and GraphicsMagick support this format.

However, both of these libraries have a large attack surface, so I'm wondering if there's a better option. Does anyone have any ideas on how to reduce the attack surface?

Thanks!

P.S. I've attached a Dockerfile for my web server.

```Dockerfile
FROM alpine:3.18

ARG VIPS_VERSION=8.14.5
ARG ALPINE_VERSION=3.18

RUN apk add xz \
    && wget https://github.com/libvips/libvips/releases/download/v${VIPS_VERSION}/vips-${VIPS_VERSION}.tar.xz \
    && tar -xf vips-${VIPS_VERSION}.tar.xz \
    && mv vips-${VIPS_VERSION} /tmp \
    && apk add \

    #######################
    #      Deps meson     #
    #######################

    meson \
    build-base \
    gobject-introspection-dev \

    #######################
    #     Deps libvips    #
    #######################

    vips-dev \
    fftw-dev \

    # Imagemagick layer
    imagemagick \

    # Portability layer
    glib-dev \

    # Rich I/O support: https://pagure.io/libaio
    libaio-dev \
    libgsf-dev \

    # XML parser: https://libexpat.github.io/
    expat-dev expat \

    # Exif support for JPEG files: https://libexif.github.io/
    libexif-dev \

    # WebP support: https://developers.google.com/speed/webp/docs/api
    libwebp-dev \

    # EXR support: https://openexr.com/en/latest/
    openexr-dev \

    # JPEG support: http://libjpeg.sourceforge.net/
    libjpeg-turbo-dev \

    # TFF support: http://www.libtiff.org/
    tiff-dev \

    # PDF support: https://poppler.freedesktop.org/
    poppler-glib \

    # SVG support: https://gitlab.gnome.org/GNOME/librsvg
    librsvg-dev \

    # Pngquant support: https://pngquant.org/
    libimagequant-dev \

    # HEIF/HEIC support: https://github.com/strukturag/libheif
    libheif-dev \

    # JXL support
    libjxl-dev \

    #######################
    #    Build libvips    #
    #######################

    && cd /tmp/vips-${VIPS_VERSION} \
    && meson setup build-dir \
    && cd build-dir \
    && meson compile \
    && meson install \

    #######################
    #      Clean up       #
    #######################

    && apk del build-base gobject-introspection-dev \
    && rm -rf /var/cache/apk/* \
    && rm -rf /tmp/*
```

**Comment by lovell** [comment]
Did you see https://imagemagick.org/script/security-policy.php ?

**Reply by michielswaanen** [reply]
Hi @lovell 

Thanks again for your quick reply. I did read the policy, and I was wondering if there are any other tools that I should consider. Additionally, the libvips documentation says that I can use GraphicsMagick, but it doesn't seem to detect it in Alpine when I add the `graphicsmagick` and `graphicsmagick-dev` libraries. Any thoughts?

**Reply by lovell** [reply]
Perhaps use Ghostscript directly, without libvips integration?

**Reply by michielswaanen** [reply]
Will take a look at that!

**Reply by jcupitt** [reply]
You need to be a bit cautious with ghostscript -- it's AGPL, so if you use it, you must opensource your whole application. ImageMagick and GraphicsMagick use ghostscript for EPS indirectly, so they have the same problem.

https://www.ghostscript.com/licensing/

**Reply by michielswaanen** [reply]
Thanks @jcupitt for notifying me about that, so I'm out of options then? 😅

**Reply by jcupitt** [reply]
I think if you want to use GhostScript commercially, you need to pay Artifex for a commercial licence. I've no idea what their terms are though.

**Reply by michielswaanen** [reply]
For now, I think I will use a third-party conversion service. Security is of uttermost importance 💯
